### DISCUSSION TITLE
Disabling Retry Feature in MQTT over QUIC Broker with EMQX 5.1.3 on Ubuntu 20.04

### DISCUSSION THREAD

**Discussion by h1tm-dev** [emqx/emqx#12122] [Q&A] [answered]
Hello, emqx community members,
 
I am operating an MQTT over QUIC broker using emqx. Currently, it seems that the retry feature is automatically enabled during QUIC session establishment. I am looking for a way to disable this retry feature.
 
Question:
 
- Is there a way to disable the retry feature when operating an MQTT over QUIC broker with emqx?
 
What I've tried:
 
- I have researched emqx documentation and checked configuration files, but I couldn't find any information regarding retry. I am aware that msquic (https://github.com/microsoft/msquic), which quic (https://github.com/emqx/quic) wraps around, may have conditions related to retry.
 
Environment Information:
 
- Ubuntu 20.04
- emqx 5.1.3
 
Any information would be greatly appreciated, so if you can provide advice or details on how to disable the retry feature, I would be grateful. Thank you in advance for your assistance.

**Comment by qzhuyan** [comment]
could you be more specific what do you mean on "retry feature"?

**Reply by h1tm-dev** [reply]
Dear qzhuyan,

I am reaching out to inquire about a specific aspect of QUIC's functionality, referred to as the "retry feature."

This pertains to the features described in RFC9000, particularly in sections 8.1.2. Address Validation Using Retry Packets and 17.2.5.1. Sending a Retry Packet, accessible at [RFC9000 Section on Retry Packet](https://datatracker.ietf.org/doc/html/rfc9000#name-retry-packet).

In emqx project, I am assuming that the implementation of this feature is optional and is governed by settings similar to those used in msquic, as per the configurations detailed at [msquic Deployment Guide](https://github.com/microsoft/msquic/blob/f327c6216e36ea5ef324d192ac3a699308164949/docs/Deployment.md).

I am interested in understanding whether it is possible to modify the settings on the emqx side to either always prevent the broker from returning retry packets, or conversely, to ensure they are always returned under any conditions. Furthermore, if this functionality is currently not available, I would like to know if there are any plans to incorporate it in the future.

Sincerely,

**Reply by qzhuyan** [reply]
thanks for the info.

we have hidden config (retry_memory_limit) in emqx quic listener that you can try out.
https://github.com/emqx/emqx/blob/1ab00777183aa8882245b2058d602c5af32578de/apps/emqx/src/emqx_schema.erl#L913
you could use env override like
```
EMQX_LISTENERS__QUIC__DEFAULT__retry_memory_limit=65
```

the meaning of the value could be found at:
https://github.com/microsoft/msquic/blob/f327c6216e36ea5ef324d192ac3a699308164949/docs/Settings.md?plain=1#L50

**Reply by h1tm-dev** [reply]
Thank you for your assistance. I have an additional question regarding the configuration settings.

Specifically, if I want to always return handshake packets, should I set the configuration as follows?

```erlang
{"retry_memory_limit",
    quic_lowlevel_settings_uint(
        65535, 65535,  % Do not return retries in any state
        ?DESC(fields_mqtt_quic_listener_retry_memory_limit)
    )},

```

Conversely, if I want to always return retry packets, should I configure it like this?

```erlang
{"retry_memory_limit",
    quic_lowlevel_settings_uint(
        0, 0,  % Return retries in all states
        ?DESC(fields_mqtt_quic_listener_retry_memory_limit)
    )},

```

I have made these changes as instructed, in the **`emqx/apps/emqx/src/emqx_schema.erl`** file, Line 913 in 1ab0077 at the **`{"retry_memory_limit",`** point. Is my understanding correct?

**Reply by qzhuyan** [reply]
You don't need to change the code, you could just change the configuration via envvar 'EMQX_LISTENERS__QUIC__DEFAULT__retry_memory_limit' or via the config file.

Ref: 
https://www.emqx.io/docs/en/latest/configuration/configuration.html

**Reply by h1tm-dev** [reply, marked as answer]
Thank you for your response.

Is the following configuration appropriate?

```bash

node {
  name = "emqx@127.0.0.1"
  cookie = "emqxsecretcookie"
  data_dir = "data"
}

cluster {
  name = emqxcl
  discovery_strategy = manual
}

dashboard {
    listeners.http {
        bind = 18083
    }
}

listeners.quic.default {
  enabled = true
  bind = "0.0.0.0:14567"
  keyfile = "etc/certs/key.pem"
  certfile = "etc/certs/cert.pem"
  # Add this to chage "retry_memory_limit"
  retry_memory_limit = 0
}
```

**Reply by qzhuyan** [reply]
yes. LGTM

**Reply by h1tm-dev** [reply]
Thank you. I will try!
